Multi-bid support #1715

pm-nilesh-chate opened the pull request:
Add Multi-Bid Support: https://github.com/prebid/prebid-server/issues/1715

bsardo commented:
@pm-nilesh-chate thanks for your patience. We're ready to start reviewing this but if you could resolve the merge conflicts first that'd be great.

pm-nilesh-chate commented:
> @pm-nilesh-chate thanks for your patience. We're ready to start reviewing this but if you could resolve the merge conflicts first that'd be great.

@bsardo done

AlexBVolcy commented on exchange/exchange.go:
```diff
@@ -451,18 +465,25 @@ func recordImpMetrics(bidRequest *openrtb2.BidRequest, metricsEngine metrics.Met
+				m, ok := multiBid[bidder.String()]
+				if i > 0 && (!ok || m.TargetBidderCodePrefix == "") {
+					// apply targeting keys to only first bid if multibid is not enabled for this bidders
+					break
```
This line isn't covered by tests, could you add a test to cover this case?

pm-nilesh-chate replied:
added test cases under `TestApplyDealSupportMultiBid()`

AlexBVolcy commented on exchange/multibid.go:
```diff
@@ -0,0 +1,70 @@
+}
+
+// Get multi-bid limit for this bidder
+func (mb *ExtMultiBidMap) GetMaxBids(bidder string) int {
+	if maxBid, ok := (*mb)[bidder]; ok {
+		return *maxBid.MaxBids
+	}
+	return DefaultBidLimit
```
This function isn't covered by tests

pm-nilesh-chate replied:
added test cases under `TestExtMultiBidMap_GetMaxBids()`

AlexBVolcy commented on exchange/multibid.go:
```diff
@@ -0,0 +1,70 @@
+type ExtMultiBidMap map[string]*openrtb_ext.ExtMultiBid
+
+// Validate and add multiBid value
+func (mb *ExtMultiBidMap) Add(multiBid *openrtb_ext.ExtMultiBid) []error {
```
Each line in this function where `errs` gets appended to isn't covered by tests

pm-nilesh-chate replied:
added test cases under `TestExtMultiBidMap_Add()`

AlexBVolcy commented on exchange/utils.go:
```diff
@@ -362,6 +363,33 @@ func buildRequestExtAlternateBidderCodes(bidder string, accABC *openrtb_ext.ExtA
 }
 
+func buildRequestExtMultiBid(adapter string, reqMultiBid []*openrtb_ext.ExtMultiBid, adapterABC *openrtb_ext.ExtAlternateBidderCodes) []*openrtb_ext.ExtMultiBid {
+	// TODO: send multiBid to adapters under adapterABC
```
Is this TODO comment still needed?

pm-nilesh-chate replied:
added code to address extra-bid feature with multibid along with its test cases

AlexBVolcy commented on exchange/exchange.go:
```diff
@@ -451,18 +465,25 @@ func recordImpMetrics(bidRequest *openrtb2.BidRequest, metricsEngine metrics.Met
+		for bidder, topBidsPerBidder := range topBidsPerImp {
+			for i, topBid := range topBidsPerBidder {
+				m, ok := multiBid[bidder.String()]
+				if i > 0 && (!ok || m.TargetBidderCodePrefix == "") {
```
Is there a way to restructure this if statement to make what's happening here more clear? Perhaps some variable re-naming?

pm-nilesh-chate replied:
Added comment to explain, let me know if that helps.
Here we want to all the loop as is for the first iteration (existing behaviour), but for 2nd and following iterations we want to proceed only if multibid feature is enabled for the particular bidder

AlexBVolcy commented on exchange/targeting.go:
```diff
@@ -19,7 +20,7 @@ const MaxKeyLength = 20
 	priceGranularity  openrtb_ext.PriceGranularity
 	includeWinners    bool
-	includeBidderKeys bool
+	includeBidderKeys bool // need to be true
```
I think this comment can be deleted

pm-nilesh-chate replied:
done

pm-nilesh-chate commented:
Hi @AlexBVolcy, I've addressed the comments, could you please review the changes.

@mansinahar not sure how you go unassigned while asking for re-review, could you please review the changes

mansinahar commented:
> @mansinahar not sure how you go unassigned while asking for re-review, could you please review the changes

@pm-nilesh-chate Not sure what happened there but yeah will take a look. Thanks for the updates!

mansinahar commented on exchange/exchange.go:
```diff
@@ -350,12 +350,15 @@ func (e *exchange) HoldAuction(ctx context.Context, r AuctionRequest, debugLog *
 		r.HookExecutor.ExecuteAllProcessedBidResponsesStage(adapterBids)
 
 		if targData != nil {
+			multiBid, multiBidErrs := getExtMultiBidData(requestExt)
```
This function creates a map of multi-bids and while doing so, we're validating the multibid config passed in the request. IMO, any kind of validation should be done at the request processing layer rather than the response processing layer. I think a better place to add this validation might be in `ValidateRequestExt` function in `openrtb2/auction.go`

AlexBVolcy replied:
Yeah I think it makes sense to have this validation be done at the request layer

pm-nilesh-chate replied:
Done, moved the validation of incoming multibid ext to `ValidateRequestExt` and updated the downstream multibid ext to avoid rework.

mansinahar commented on exchange/targeting.go:
```diff
@@ -35,49 +36,80 @@ type targetData struct {
-			}
+			for i, topBid := range topBidsPerBidder {
+				if accountMaxBids != 0 && i >= accountMaxBids {
+					topBid.Bid = nil
```
Looks like we're enforcing the max bids limit in this function. Doing this in `setTargeting` is kinda misleading IMHO. We should have a separate function where the bid limit is enforced. It might be slightly inefficient but it will make the code easier to read and understand. 

Also, at this point in the code we have already done some bid processing like setting up even tracking info, doing currency conversions, etc. Is there a way we can move this logic upstream so that we get rid of the bids earlier in the process and don't avoid processing extra bids that might be potentially discarded?

pm-nilesh-chate replied:
We can certainly isolate this logic, but it should only be done after the auction has ended. We would want to eliminate only the "least wanted" or "losing bids.

related to https://github.com/prebid/prebid-server/pull/2468#discussion_r1099836817

pm-nilesh-chate commented on exchange/auction.go:
```diff
@@ -107,29 +108,36 @@ func (d *DebugLog) PutDebugLogError(cache prebid_cache_client.Client, timeout in
-					if !ok || cpm > bestSoFar.Bid.Price {
-						bidMap[bidderName] = bid
-					}
+					bidMap[bidderName] = append(bidMap[bidderName], bid)
```
For multi-bid targeting, we need to process bids per adapter grouped by `ImpID`, hence used this original code to avoid rebuilding the same map in targeting.go's `setTargeting()`

pm-nilesh-chate commented on config/accounts.go:
```diff
@@ -36,6 +36,7 @@ type Account struct {
 	AlternateBidderCodes    *openrtb_ext.ExtAlternateBidderCodes `mapstructure:"alternatebiddercodes" json:"alternatebiddercodes"`
 	Hooks                   AccountHooks                         `mapstructure:"hooks" json:"hooks"`
 	Validations             Validations                          `mapstructure:"validations" json:"validations"`
+	DefaultBidLimitMin      int                                  `mapstructure:"default_bid_limit_min" json:"default_bid_limit_min"`
```
Picked the name from comment https://github.com/prebid/prebid-server/pull/2254#issuecomment-1188511513. Will update PR with new name after confirmation from tagged folks.

cc @bretg

AlexBVolcy replied:
I agree with @mansinahar about the name changing to `DefaultBidsLimit `

pm-nilesh-chate replied:
done

pm-nilesh-chate commented on exchange/multibid.go:
```diff
@@ -0,0 +1,73 @@
+const DefaultBidLimit = 1
+const MaxBidLimit = 9
+
+type ExtMultiBidMap map[string]*openrtb_ext.ExtMultiBid
```
This config would be ready only. Updated the code accordingly.

pm-nilesh-chate commented on exchange/exchange.go:
```diff
@@ -374,8 +377,19 @@ func (e *exchange) HoldAuction(ctx context.Context, r AuctionRequest, debugLog *
-			targData.setTargeting(auc, r.BidRequestWrapper.BidRequest.App != nil, bidCategory, r.Account.TruncateTargetAttribute)
+			targData.setTargeting(auc, r.BidRequestWrapper.BidRequest.App != nil, bidCategory, r.Account.TruncateTargetAttribute, multiBid, r.Account.DefaultBidLimitMin)
 
+			if r.Account.DefaultBidLimitMin != 0 {
```
Moved this multi-bid logic and the one from `setTargeting()` to new `validateAndUpdateMultiBid()`

The `DefaultBidLimitMin` from https://github.com/prebid/prebid-server/pull/2254#issuecomment-1188511513 sets a strict limit on the number of bids each adapter can send. The `adapterBids` stores these bids in the order they were received from the adapters, while the auction logic sorts the bids in `auction.winningBidsByBidder`. As a result, the sequence of bids in the exchange.go's `adapterBids` and the `auction.winningBidsByBidder` will differ. In targeting.go, we set bids to be dropped as nil to remove them from the `adapterBids` slice in exchange.go. This also ensures that there is no memory leak by explicitly marking dropped bids and removing any unused references in the slice.

pm-nilesh-chate commented on exchange/exchange.go:
```diff
@@ -451,18 +465,25 @@ func recordImpMetrics(bidRequest *openrtb2.BidRequest, metricsEngine metrics.Met
+		for bidder, topBidsPerBidder := range topBidsPerImp {
+			for i, topBid := range topBidsPerBidder {
+				bidderMultiBid, ok := multiBid[bidder.String()]
+				if i > 0 && (!ok || bidderMultiBid.TargetBidderCodePrefix == "") {
```
Targeting keys for more than 1 bid will be updated only if multi-bid config is specified. `default_bid_limit_min` was introduced only to pair it with multi-bid feature as a hard limit on bid count per adapter to have a Java-PBS like behaviour and keep the existing behaviour of Go-PBS as is which is to forward all the bids as is to caller. Details at https://github.com/prebid/prebid-server/pull/2254#issuecomment-1188511513.

pm-nilesh-chate commented on exchange/auction.go:
```diff
@@ -150,10 +158,12 @@ func isNewWinningBid(bid, wbid *openrtb2.Bid, preferDeals bool) bool {
 
 func (a *auction) setRoundedPrices(priceGranularity openrtb_ext.PriceGranularity) {
-	roundedPrices := make(map[*entities.PbsOrtbBid]string, 5*len(a.winningBids))
+	roundedPrices := make(map[*entities.PbsOrtbBid]string, 5*len(a.winningBids)) // TODO
```
removed

pm-nilesh-chate commented on exchange/targeting_test.go:
```diff
@@ -610,16 +611,16 @@ var TargetingTests []TargetingTestData = []TargetingTestData{
 						Bid:     bid084,
 						BidType: openrtb_ext.BidTypeBanner,
-					},
+					}},
```
- Multi-bid scenarios are covered as part of `multi-bids.json`, `multi-bids-with-extra-bids.json` and `multi-bid-default-bid-limit-min.json` in `TestExchangeJSON()`.
- Updated `TestSetTargeting()` as well with multi-bid scenarios.

pm-nilesh-chate commented on exchange/multibid.go:
```diff
@@ -0,0 +1,73 @@
+const DefaultBidLimit = 1
+const MaxBidLimit = 9
+
+type ExtMultiBidMap map[string]*openrtb_ext.ExtMultiBid
```
Done, moved the validation of incoming multibid ext to `ValidateRequestExt` and updated the downstream multibid ext to avoid rework.

pm-nilesh-chate commented on exchange/multibid.go:
```diff
@@ -0,0 +1,73 @@
+			}
+
+			if multiBid.TargetBidderCodePrefix != "" {
+				errs = append(errs, fmt.Errorf("ignoring targetbiddercodeprefix for %v", multiBid))
```
Done, added a new warning code and updated the response warnings as part of multibid ext validation in `ValidateRequestExt`.

pm-nilesh-chate commented:
Hi @mansinahar, I've addressed the comments. Could you please review the same. I've resolved the addressed conversations to keep only active ones open. Please reopen if any of them needs to be followed up.

cc @AlexBVolcy

AlexBVolcy commented on exchange/auction.go:
```diff
@@ -149,11 +148,50 @@ func isNewWinningBid(bid, wbid *openrtb2.Bid, preferDeals bool) bool {
+				return isNewWinningBid(topBids[i].Bid, topBids[j].Bid, preferDeals)
+			})
+
+			// assert hard limit on bids count per imp, per adapter. ()
```
Nitpick: Can you delete the parentheses at the end of this comment

pm-nilesh-chate replied:
done

AlexBVolcy commented on exchange/auction.go:
```diff
@@ -149,11 +148,50 @@ func isNewWinningBid(bid, wbid *openrtb2.Bid, preferDeals bool) bool {
+				for i := accountDefaultBidLimit; i < len(topBids); i++ {
+					topBids[i].Bid = nil
+					topBids[i] = nil
+					bidsDropped = true // mark drop-candidate bid and remove its reference (avoid mem leak)
```
Nitpick: I think this comment can be removed.

pm-nilesh-chate replied:
done

AlexBVolcy commented on openrtb_ext/request.go:
```diff
@@ -72,6 +73,9 @@ type ExtRequestPrebid struct {
 	Trace string `json:"trace,omitempty"`
+
+	MultiBid    []*ExtMultiBid         `json:"multibid,omitempty"`
+	MultibidMap map[string]ExtMultiBid `json:"-"`
```
Nitpick: Can we make the 'b' in bid uppercase `MultiBidMap`

pm-nilesh-chate replied:
done

AlexBVolcy commented on openrtb_ext/bid.go:
```diff
@@ -19,6 +19,7 @@ type ExtBidPrebid struct {
 	DealTierSatisfied bool                `json:"dealtiersatisfied,omitempty"`
 	Meta              *ExtBidPrebidMeta   `json:"meta,omitempty"`
 	Targeting         map[string]string   `json:"targeting,omitempty"`
+	Targetbiddercode  string              `json:"targetbiddercode,omitempty"`
```
Can this be renamed with capitalizations to be `TargetBidderCode`

pm-nilesh-chate replied:
done

AlexBVolcy commented on exchange/targeting.go:
```diff
@@ -35,49 +36,66 @@ type targetData struct {
-				targData.addKeys(targets, openrtb_ext.HbFormatKey, string(topBidPerBidder.BidType), bidderName, isOverallWinner, truncateTargetAttr)
-			}
+			for i, topBid := range topBidsPerBidder {
+				if i == maxBids || // limit targeting keys to maxBids (default 1 bid),
```
Can this be re-written so that both conditions are on the same line, and then you can have a comment above the if statement?

`// Comment Here`
`if i == maxBids || (i == 1 && bidderCodePrefix == ""`

pm-nilesh-chate replied:
done

AlexBVolcy commented on openrtb_ext/multibid.go:
```diff
@@ -0,0 +1,89 @@
+
+	var validationErrs, errs []error
+	var validatedMultiBids, newMultiBids []*ExtMultiBid //returning slice instead of map to keep the downstream req.Ext payload consistent
+	multibidMap := make(map[string]struct{})            // map is needed temporarily for validate of duplicate entries, etc.
```
Nitpick: can this be renamed to `multiBidMap` and I think this comment can be removed.

pm-nilesh-chate replied:
done

AlexBVolcy commented on openrtb_ext/multibid.go:
```diff
@@ -0,0 +1,89 @@
+	}
+
+	var validatedMultiBids []*ExtMultiBid
+	// Prefer Bidder over []Bidders
```
I think this comment can be deleted

pm-nilesh-chate replied:
done

pm-nilesh-chate commented on exchange/targeting.go:
```diff
@@ -35,49 +36,67 @@ type targetData struct {
 // it's ok if those stay in the auction. For now, this method implements a very naive cache strategy.
 // In the future, we should implement a more clever retry & backoff strategy to balance the success rate & performance.
-func (targData *targetData) setTargeting(auc *auction, isApp bool, categoryMapping map[string]string, truncateTargetAttr *int) {
+func (targData *targetData) setTargeting(auc *auction, isApp bool, categoryMapping map[string]string, truncateTargetAttr *int, multiBidMap map[string]openrtb_ext.ExtMultiBid, accountMaxBids int) {
```
removed.

pm-nilesh-chate commented on config/accounts.go:
```diff
@@ -36,6 +36,7 @@ type Account struct {
 	AlternateBidderCodes    *openrtb_ext.ExtAlternateBidderCodes `mapstructure:"alternatebiddercodes" json:"alternatebiddercodes"`
 	Hooks                   AccountHooks                         `mapstructure:"hooks" json:"hooks"`
 	Validations             Validations                          `mapstructure:"validations" json:"validations"`
+	DefaultBidLimit         int                                  `mapstructure:"default_bid_limit" json:"default_bid_limit"`
```
This option was introduced as a solution to a difference in behavior between PBS-Go and PBS-Java when multiple bids are returned for the same impression. PBS-Java drops any extra bids beyond a certain hard limit, while PBS-Go forwards all the bids without processing or dropping them. https://github.com/prebid/prebid-server/issues/1715#issuecomment-1148681203

To address this difference in behavior and to provide more control to hosts or accounts, the "DefaultBidLimit" field was introduced. This allows the host or account to set a hard limit on the number of bids they want to receive, and any bids beyond that limit will be dropped. This feature allows PBS-Go users to maintain the existing behavior while also having the option to implement a multibid feature similar to that of PBS-Java.

AlexBVolcy commented on exchange/exchangetest/multi-bid-default-bid-limit-min.json:
```diff
@@ -0,0 +1,707 @@
+{
```
Because of how we updated the naming of `DefaultBidLimitMin` --> `DefaultBidLimit`

Could you remove the -min from the .json filename?

pm-nilesh-chate replied:
done

AlexBVolcy commented on exchange/exchangetest/multi-bids-with-extra-bids.json:
```diff
@@ -0,0 +1,489 @@
+{
```
Could you add another JSON exchange test where the number of bids returned by a seat for a particular impID is greater than the maxBids value associated with that seat? That way we can see that bids are properly including/excluding targeting data.

pm-nilesh-chate replied:
`multi-bids.json` handles this case where seat `pubmatic` gets 3 bids for impID `my-imp-id` and `maxBids` in request is 2. Hence, in the response we have targeting applied only for the first 2 bids and the last bid with id `losing-bid` does not have any targeting.

Am I missing anything here?

pm-nilesh-chate commented on exchange/auction.go:
```diff
@@ -149,11 +148,50 @@ func isNewWinningBid(bid, wbid *openrtb2.Bid, preferDeals bool) bool {
 }
 
+func (a *auction) validateAndUpdateMultiBid(adapterBids map[openrtb_ext.BidderName]*entities.PbsOrtbSeatBid, preferDeals bool, accountDefaultBidLimit int) {
+	bidsDropped := false
```
done

pm-nilesh-chate commented on exchange/auction_test.go:
```diff
@@ -537,6 +585,257 @@ func TestNewAuction(t *testing.T) {
 
 }
 
+func Test_auction_validateAndUpdateMultiBid(t *testing.T) {
```
done

pm-nilesh-chate commented on exchange/auction.go:
```diff
@@ -149,11 +148,50 @@ func isNewWinningBid(bid, wbid *openrtb2.Bid, preferDeals bool) bool {
+					bidsDropped = true
+				}
+
+				topBidsPerBidder[bidder] = topBids[:accountDefaultBidLimit]
```
I see this already being discussed. Leaving the link for reference https://github.com/prebid/prebid-server/issues/1715#issuecomment-1447047209

VeronikaSolovei9 reviewed (dismissed):
We confirmed expected behavior and config names with Bret. Local testing looks good. Thank you for your contribution!

hhhjort commented on exchange/exchange_test.go:
```diff
@@ -3494,29 +3519,213 @@ func TestApplyDealSupport(t *testing.T) {
 	}
 }
 
+var getIntPtr = func(m int) *int { return &m }
```
We now have `/util/ptrutil/ToPtr()` so you don't need to define your own copy here.

pm-nilesh-chate replied:
done, thanks :)
